### ISSUE TITLE
Eureka Client getting connection refused on EC2

### ISSUE THREAD

**Issue by cloudnt** [spring-cloud/spring-cloud-netflix#2224]
Hi,

I am trying to run simple spring boot application with eureka discovery and eureka client registering to the discovery. Both I am starting on 2 different EC2 instances. My discovery server is starting fine but client is throwing "Connection refused" error while registering to Eureka Discovery Server. I've looked different posts and Eureka Github page but at this time I am not worried about autoscaling/load balancing eureka. that's why running only one eureka server. 

Below are my bootstrap.yml for Eureka discovery and Eureka Client :

**### Discovery server bootstrap.yml:**
```
eureka:
  client:
    register-with-eureka: false
    fetch-registry: false
    service-url:
      defaultZone: ${eureka.url:http://localhost:8761/}
  instance:
    hostname: ${eureka.host:localhost}
    data-center-info:
      datacenter: Amazon
  server:
    enable-self-preservation: false
    wait-time-in-ms-when-sync-empty: 0
spring:
  profiles: discovery-server-aws
server:
  port: ${server.port:8761}
```
**### Eureka Client bootstrap.yml:**
```
spring:
  application:
    name: nt-gateway-aws
server:
  port: ${server.port:9001}
  context-path: /gw
eureka:
  instance:
    data-center-info:
      datacenter: Amazon
  client:
    register-with-eureka: true
    fetch-registry: true
    service-url:
      default-zone: ${eureka.url:http://localhost:8761/eureka/}
```
Everything works locally perfectly all right but when I deploy in AWS EC2, it fails. And below are how I am starting discovery server and eureka client :

**### Discovery server startup :**
```java -jar -Dserver.port=8761 -Deureka.url=http://EC2_PUBLIC_DOMAIN:8761/ -Deureka.host=EC2_PUBLIC_DOMAIN -Deureka.datacenter=Amazon ./discovery-server-aws-1.0.jar```

**### Eureka Client startup:**
```java -jar -Dserver.port=9001 -Deureka.url=http://EC2_PRIVATE_IP:8761/eureka/ -Deureka.datacenter=Amazon ./nt-gateway-aws-1.0.jar```

Any help is appreciated.

Thank you

**Comment by spencergibb** [comment]
Can you please provide more information, such as the stack trace? "it fails" isn't very descriptive.

**Comment by cloudnt** [comment]
Here's the stacktrace:

com.sun.jersey.api.client.ClientHandlerException: java.net.ConnectException: Connection refused (Connection refused)
	at com.sun.jersey.client.apache4.ApacheHttpClient4Handler.handle(ApacheHttpClient4Handler.java:187) ~[jersey-apache-client4-1.19.1.jar!/:1.19.1]
	at com.sun.jersey.api.client.filter.GZIPContentEncodingFilter.handle(GZIPContentEncodingFilter.java:123) ~[jersey-client-1.19.1.jar!/:1.19.1]
	at com.netflix.discovery.EurekaIdentityHeaderFilter.handle(EurekaIdentityHeaderFilter.java:27) ~[eureka-client-1.6.2.jar!/:1.6.2]
	at com.sun.jersey.api.client.Client.handle(Client.java:652) ~[jersey-client-1.19.1.jar!/:1.19.1]
	at com.sun.jersey.api.client.WebResource.handle(WebResource.java:682) ~[jersey-client-1.19.1.jar!/:1.19.1]
	at com.sun.jersey.api.client.WebResource.access$200(WebResource.java:74) ~[jersey-client-1.19.1.jar!/:1.19.1]
	at com.sun.jersey.api.client.WebResource$Builder.delete(WebResource.java:591) ~[jersey-client-1.19.1.jar!/:1.19.1]
	at com.netflix.discovery.shared.transport.jersey.AbstractJerseyEurekaHttpClient.cancel(AbstractJerseyEurekaHttpClient.java:76) ~[eureka-client-1.6.2.jar!/:1.6.2]
	at com.netflix.discovery.shared.transport.decorator.EurekaHttpClientDecorator$2.execute(EurekaHttpClientDecorator.java:74) [eureka-client-1.6.2.jar!/:1.6.2]
	at com.netflix.discovery.shared.transport.decorator.MetricsCollectingEurekaHttpClient.execute(MetricsCollectingEurekaHttpClient.java:73) ~[eureka-client-1.6.2.jar!/:1.6.2]
	at com.netflix.discovery.shared.transport.decorator.EurekaHttpClientDecorator.cancel(EurekaHttpClientDecorator.java:71) [eureka-client-1.6.2.jar!/:1.6.2]
	at com.netflix.discovery.shared.transport.decorator.EurekaHttpClientDecorator$2.execute(EurekaHttpClientDecorator.java:74) [eureka-client-1.6.2.jar!/:1.6.2]
	at com.netflix.discovery.shared.transport.decorator.RedirectingEurekaHttpClient.executeOnNewServer(RedirectingEurekaHttpClient.java:118) ~[eureka-client-1.6.2.jar!/:1.6.2]
	at com.netflix.discovery.shared.transport.decorator.RedirectingEurekaHttpClient.execute(RedirectingEurekaHttpClient.java:79) ~[eureka-client-1.6.2.jar!/:1.6.2]
	at com.netflix.discovery.shared.transport.decorator.EurekaHttpClientDecorator.cancel(EurekaHttpClientDecorator.java:71) [eureka-client-1.6.2.jar!/:1.6.2]
	at com.netflix.discovery.shared.transport.decorator.EurekaHttpClientDecorator$2.execute(EurekaHttpClientDecorator.java:74) [eureka-client-1.6.2.jar!/:1.6.2]
	at com.netflix.discovery.shared.transport.decorator.RetryableEurekaHttpClient.execute(RetryableEurekaHttpClient.java:119) [eureka-client-1.6.2.jar!/:1.6.2]
	at com.netflix.discovery.shared.transport.decorator.EurekaHttpClientDecorator.cancel(EurekaHttpClientDecorator.java:71) [eureka-client-1.6.2.jar!/:1.6.2]
	at com.netflix.discovery.shared.transport.decorator.EurekaHttpClientDecorator$2.execute(EurekaHttpClientDecorator.java:74) [eureka-client-1.6.2.jar!/:1.6.2]
	at com.netflix.discovery.shared.transport.decorator.SessionedEurekaHttpClient.execute(SessionedEurekaHttpClient.java:77) [eureka-client-1.6.2.jar!/:1.6.2]
	at com.netflix.discovery.shared.transport.decorator.EurekaHttpClientDecorator.cancel(EurekaHttpClientDecorator.java:71) [eureka-client-1.6.2.jar!/:1.6.2]
	at com.netflix.discovery.DiscoveryClient.unregister(DiscoveryClient.java:885) [eureka-client-1.6.2.jar!/:1.6.2]
	at com.netflix.discovery.DiscoveryClient.shutdown(DiscoveryClient.java:863) [eureka-client-1.6.2.jar!/:1.6.2]
	at com.netflix.discovery.DiscoveryClient$$FastClassBySpringCGLIB$$a84c8cb4.invoke(<generated>) [eureka-client-1.6.2.jar!/:1.6.2]
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:204) [spring-core-4.3.10.RELEASE.jar!/:4.3.10.RELEASE]
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:738) [spring-aop-4.3.10.RELEASE.jar!/:4.3.10.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:157) [spring-aop-4.3.10.RELEASE.jar!/:4.3.10.RELEASE]
	at org.springframework.cloud.context.config.StandardBeanLifecycleDecorator$2.invoke(StandardBeanLifecycleDecorator.java:85) [spring-cloud-context-1.2.3.RELEASE.jar!/:1.2.3.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179) [spring-aop-4.3.10.RELEASE.jar!/:4.3.10.RELEASE]
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:673) [spring-aop-4.3.10.RELEASE.jar!/:4.3.10.RELEASE]
	at org.springframework.cloud.netflix.eureka.CloudEurekaClient$$EnhancerBySpringCGLIB$$2e8b8520.shutdown(<generated>) [spring-cloud-netflix-eureka-client-1.3.4.RELEASE.jar!/:1.3.4.RELEASE]
	at org.springframework.cloud.netflix.eureka.serviceregistry.EurekaServiceRegistry.deregister(EurekaServiceRegistry.java:71) [spring-cloud-netflix-eureka-client-1.3.4.RELEASE.jar!/:1.3.4.RELEASE]
	at org.springframework.cloud.netflix.eureka.serviceregistry.EurekaAutoServiceRegistration.stop(EurekaAutoServiceRegistration.java:82) [spring-cloud-netflix-eureka-client-1.3.4.RELEASE.jar!/:1.3.4.RELEASE]
	at org.springframework.cloud.netflix.eureka.serviceregistry.EurekaAutoServiceRegistration.onApplicationEvent(EurekaAutoServiceRegistration.java:126) [spring-cloud-netflix-eureka-client-1.3.4.RELEASE.jar!/:1.3.4.RELEASE]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_144]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:1.8.0_144]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_144]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[na:1.8.0_144]
	at org.springframework.context.event.ApplicationListenerMethodAdapter.doInvoke(ApplicationListenerMethodAdapter.java:256) [spring-context-4.3.10.RELEASE.jar!/:4.3.10.RELEASE]
	at org.springframework.context.event.ApplicationListenerMethodAdapter.processEvent(ApplicationListenerMethodAdapter.java:177) [spring-context-4.3.10.RELEASE.jar!/:4.3.10.RELEASE]
	at org.springframework.context.event.ApplicationListenerMethodAdapter.onApplicationEvent(ApplicationListenerMethodAdapter.java:140) [spring-context-4.3.10.RELEASE.jar!/:4.3.10.RELEASE]
	at org.springframework.context.event.SimpleApplicationEventMulticaster.invokeListener(SimpleApplicationEventMulticaster.java:167) [spring-context-4.3.10.RELEASE.jar!/:4.3.10.RELEASE]
	at org.springframework.context.event.SimpleApplicationEventMulticaster.multicastEvent(SimpleApplicationEventMulticaster.java:139) [spring-context-4.3.10.RELEASE.jar!/:4.3.10.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext.publishEvent(AbstractApplicationContext.java:393) [spring-context-4.3.10.RELEASE.jar!/:4.3.10.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext.publishEvent(AbstractApplicationContext.java:347) [spring-context-4.3.10.RELEASE.jar!/:4.3.10.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext.doClose(AbstractApplicationContext.java:991) [spring-context-4.3.10.RELEASE.jar!/:4.3.10.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext$2.run(AbstractApplicationContext.java:929) [spring-context-4.3.10.RELEASE.jar!/:4.3.10.RELEASE]
Caused by: java.net.ConnectException: Connection refused (Connection refused)
	at java.net.PlainSocketImpl.socketConnect(Native Method) ~[na:1.8.0_144]
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350) ~[na:1.8.0_144]
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206) ~[na:1.8.0_144]
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188) ~[na:1.8.0_144]
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392) ~[na:1.8.0_144]
	at java.net.Socket.connect(Socket.java:589) ~[na:1.8.0_144]
	at org.apache.http.conn.scheme.PlainSocketFactory.connectSocket(PlainSocketFactory.java:121) ~[httpclient-4.5.3.jar!/:4.5.3]
	at org.apache.http.impl.conn.DefaultClientConnectionOperator.openConnection(DefaultClientConnectionOperator.java:180) ~[httpclient-4.5.3.jar!/:4.5.3]
	at org.apache.http.impl.conn.AbstractPoolEntry.open(AbstractPoolEntry.java:144) ~[httpclient-4.5.3.jar!/:4.5.3]
	at org.apache.http.impl.conn.AbstractPooledConnAdapter.open(AbstractPooledConnAdapter.java:134) ~[httpclient-4.5.3.jar!/:4.5.3]
	at org.apache.http.impl.client.DefaultRequestDirector.tryConnect(DefaultRequestDirector.java:610) ~[httpclient-4.5.3.jar!/:4.5.3]
	at org.apache.http.impl.client.DefaultRequestDirector.execute(DefaultRequestDirector.java:445) ~[httpclient-4.5.3.jar!/:4.5.3]
	at org.apache.http.impl.client.AbstractHttpClient.doExecute(AbstractHttpClient.java:835) ~[httpclient-4.5.3.jar!/:4.5.3]
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:118) ~[httpclient-4.5.3.jar!/:4.5.3]
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:56) ~[httpclient-4.5.3.jar!/:4.5.3]
	at com.sun.jersey.client.apache4.ApacheHttpClient4Handler.handle(ApacheHttpClient4Handler.java:173) ~[jersey-apache-client4-1.19.1.jar!/:1.19.1]
	... 46 common frames omitted

2017-08-23 17:37:28.607  WARN 23422 --- [       Thread-6] c.n.d.s.t.d.RetryableEurekaHttpClient    : Request execution failed with message: java.net.ConnectException: Connection refused (Connection refused)
2017-08-23 17:37:28.607 ERROR 23422 --- [       Thread-6] com.netflix.discovery.DiscoveryClient    : DiscoveryClient_NT-GATEWAY-AWS/ip-172-31-9-97.ec2.internal:nt-gateway-aws:9001 - de-registration failedCannot execute request on any known server

com.netflix.discovery.shared.transport.TransportException: Cannot execute request on any known server

**Comment by spencergibb** [comment]
Doesn't look like the eureka client can hit the public url of the eureka server.

**Comment by spencergibb** [comment]
I'm unsure how I can help.

**Comment by cloudnt** [comment]
I actually tried putting EC2 internal IP as well but still not able to do that.  As per this url https://github.com/Netflix/eureka/wiki/Deploying-Eureka-Servers-in-EC2 I tried passing that special jvm arg to tell Eureka about datacenter=cloud.  But that didn't help either. So, stuck totally. Nowhere it says any additional security group or role to be setup for running eureka. 

So, any help from anyone would be great.

Thank you

**Comment by spencergibb** [comment]
Not sure that has anything to do with a connection refused. I'd check AWS networking rules

**Comment by cloudnt** [comment]
I actually tried running `curl http://DISCOVERY_EC2_PUBLIC_DOMAIN:8761` and it worked fine. So, that tells me that there's no connectivity issue from client EC2 host to Discovery Server EC2 host. But still starting client is throwing "Connection Refused" error.

Any insight - anyone ?

**Comment by spencergibb** [comment]
Can you show me /configprops for the client, specifically the eureka sections

**Comment by cloudnt** [comment]
it is :
```
eureka:
  instance:
    data-center-info:
      datacenter: Amazon
  client:
    register-with-eureka: true
    fetch-registry: true
    service-url:
      default-zone: ${eureka.url:http://localhost:8761/eureka/}
```

and I am passing jvm arg as 
`-Deureka.url=http://DISCOVERY_SERVER_EC2_PUBLIC_DOMAIN:8761/eureka/`

**Comment by spencergibb** [comment]
That's not the output of the `/configprops` actuator endpoint.

**Comment by cloudnt** [comment]
aah.. ok. but I don't have actuator added as dependency in my eureka client. And I am not sure even if i add that dependency if its going to help much because client is not even starting because of that error.

**Comment by spencergibb** [comment]
ok, I'm guessing your parameter isn't actually making it to your configuration. If you turn apache http client debugging on, you can see the host it's trying to connect to. My guess, localhost. Try `--eureka.url` after the `.jar` rather than `-D`. Also do `--logging.level.org.apache.http.wire=DEBUG`.

**Comment by cloudnt** [comment]
tried both option but nothing worked. even not seeing full httpclient log even after adding that log enabling jvm arg.. so, can't really tell what is the http call its making.  But i think it is reading the value from jvm arg correctly. Because i see that it is trying to start client on port 9001 which is what i m passing as -Dserver.port=9001

**Comment by spencergibb** [comment]
The app shouldn't fail if eureka can't connect. I've not had that happen to me. So you should be able to see `/configprops`

**Comment by spencergibb** [comment]
The error is actually happening on `de-registration`

**Comment by spencergibb** [comment]
What version are you using (ie what release train)?

**Comment by cloudnt** [comment]
spring.cloud.version=Dalston.SR3

and spring boot version = 1.5.6.RELEASE

**Comment by cloudnt** [comment]
Phew!!! finally able to get it working. First of all BIG THANK YOU to Spencer for right pointers on ```/configprops ```. I added actuator endpoint and found out that in my discovery server i had ```eureka.client.service-url.defaultZone```  whereas in Eureka client, I had it ```eureka.client.service-url. default-zone```.

And because of that, client was trying to connect to localhost:8761 as you pointed out. I changed client prop to match with discovery server and Voila!!!

Thanks again.
